pkg/retriever/graphsyncretriever.go

```diff
@@ -131,29 +131,55 @@ func (r *graphsyncRetrieval) RetrieveFromCandidates(candidates []types.Retrieval
 	}
 	// start retrievals
 	queryStartTime := time.Now()
-	r.waitGroup.Add(len(candidates))
-	for _, candidate := range candidates {
-		candidate := candidate
-		go func() {
-			runRetrievalCandidate(ctx, r.GraphSyncRetriever, r.request, r.Client, retrieval, queryStartTime, candidate)
-			r.waitGroup.Done()
-		}()
-	}
+	r.waitGroup.Add(1)
+	go func() {
+		defer r.waitGroup.Done()
+		for {
+			hasCandidates, candidates, err := asyncCandidates.Next(ctx)
+			if !hasCandidates || err != nil {
+				return
+			}
+			for _, candidate := range candidates {
+				candidate := candidate
+				r.waitGroup.Add(1)
+				go func() {
+					defer r.waitGroup.Done()
+					runRetrievalCandidate(ctx, r.GraphSyncRetriever, r.request, r.Client, retrieval, queryStartTime, candidate)
```

Comment:
well, that was surprisingly easy, I guess we do it all async with goroutines and channels anyway so staggered incoming isn't unreasonable; getting rid of `finishedCount` even makes `collectResults` a bit nicer